cmake/toolchain.cmake

```diff
@@ -36,3 +36,17 @@ endif()
 # Configure the toolchain based on what toolchain technology is used
 # (gcc, host-gcc etc.)
 include(${ZEPHYR_BASE}/cmake/compiler/${COMPILER}.cmake OPTIONAL)
+
+# Uniquely identify the toolchain wrt. it's capabilities.
+#
+# What we are looking for, is a signature definition that is defined
+# like this:
+
+# Toolchains with the same signature will always support the same set
+# of flags.
+
+# It is not clear how this signature should be constructed. The
+# strategy chosen is to md5sum the CC binary.
+
+file(MD5 ${CMAKE_C_COMPILER} CMAKE_C_COMPILER_MD5_SUM)
+set(TOOLCHAIN_SIGNATURE ${CMAKE_C_COMPILER_MD5_SUM})
```

Comment:
This seems like it is a "good enough" heuristic to me, but I don't know how it might fail.